lib/datalog/access.go

```diff
@@ -0,0 +1,378 @@
+// +build roletester
+
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package datalog
+
+// #cgo LDFLAGS: -L./roletester/target/release -lrole_tester
+// #include <stdio.h>
+// #include <stdlib.h>
+// typedef struct status status_t;
+// extern status_t *process_access(unsigned char *input, unsigned char *output, size_t input_len, size_t output_len);
+// extern size_t status_output_length(status_t *);
+// extern int status_error(status_t *);
+// extern void drop_status_struct(status_t *status);
+import "C"
+import (
+	"context"
+	"fmt"
+	"hash/fnv"
+	"sort"
+	"strings"
+	"unsafe"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/asciitable"
+	"github.com/gravitational/teleport/lib/auth"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gravitational/trace"
+)
+
+// NodeAccessRequest defines a request for access for a specific user, login, and node.
+type NodeAccessRequest struct {
+	Username  string
+	Login     string
+	Node      string
+	Namespace string
+}
+
+// NodeAccessResponse defines all interpreted facts from the input facts.
+type NodeAccessResponse struct {
+	// Generated output from Rust.
+	Accesses Facts
+	// Generated input for Rust.
+	facts Facts
+	// Mappings to convert from hash values to readable strings.
+	mappings        map[string]uint32
+	reverseMappings map[uint32]string
```

Comment:
@rueey Could you explain here with a comment why these mapping are needed? Looking at Rust part, all predicates are defined as `struct HasRole(u32, u32);` which I'm guessing is why you need to convert to uint32, but could we define those predicates as `HasRole(string, string)` (or Rust equivalent) so we could pass all values as it is (as literal strings) without hashing? From Rust<->Go interop perspective it shouldn't cause problems since we're marshaling to protobuf?

Reply:
That said, avoiding strings will probably be fastest for large counts of those since moving large strings across FFI boundaries will never be fast.

Reply:
Mainly just for performance reasons. Another factor was how complicated the rust code would have been if I had to do string handling, with the ownership model + lifetimes which I'm still working to grasp fully. At the time of implementation, it was difficult to implement for me in rust with crepe since there wasn't much documentation. Hashing on the Go end seemed like the best solution without compromising too much in terms of simpler implementation. But this could definitely be a future improvement if we find performance is not too bad.

Reply:
@rueey Makes sense, let's please add a comment on these two fields here then so explain that.